steps/src/gps/mod.rs

```diff
@@ -0,0 +1,227 @@
+//! Distance between coordinates.
+//!
+//! If we have two coordinates we could use Haversine formula:
+//!
+//! ```norust
+//! d=2R*sin ^ −1(√(sin^2((Φ2​−Φ1​​)/2)+cos(Φ1​)cos(Φ2​)sin^2((λ2​−λ1​​)/2)))
+//! ```
+//!
+//! where:
+//!
+//! - R – Earth's radius (R = 6371 km);
+//! - λ1, φ₁ – First point longitude and latitude coordinates;
+//! - λ2, φ₂ – Second point longitude and latitude coordinates;
+//! - d – Distance between them along Earth's surface.
+
+mod models;
+
+pub use models::*;
+
+/// Radius of Earth
+pub const R: f64 = 6371.0087714150598;
+
+const WINDOW_SIZE: usize = 2;
+const SPEED_THRESHOLD_KMPHR: f64 = 20.0;
+
+pub fn movement_from_gps(data: impl IntoIterator<Item = Gps>) -> Vec<Movement> {
+    let data = data.into_iter().collect::<Vec<_>>();
+
+    data.windows(WINDOW_SIZE)
+        .map(|this| {
+            let first = &this[0];
+            let second = &this[1];
+
+            let distance = match (first.altitude, second.altitude) {
+                (Some(altitude_1), Some(altitude_2)) => {
+                    let flat_distance = haversine(
+                        first.longitude,
+                        first.latitude,
+                        second.longitude,
+                        second.latitude,
+                    );
+
+                    (flat_distance.powi(2) + ((altitude_2 - altitude_1) / 1000.0).powi(2)).sqrt()
+                }
+                _ => haversine(
+                    first.longitude,
+                    first.latitude,
+                    second.longitude,
+                    second.latitude,
+                ),
+            };
+
+            Movement {
+                distance: Distance::from_kilometers(distance),
+                duration: second.timestamp - first.timestamp,
+            }
+        })
+        .collect()
+}
+
+/// Calculate number of steps using data from GPS.
+/// This method filters out movement if it was above threshold e.g. bicycling or driving a car.
+///
+/// # Params
+/// - data - gps data which sorted by timestamp in asc order
+/// - height - height of person in meters
+pub fn steps_from_gps(data: impl IntoIterator<Item = Gps>, height: f64) -> f64 {
+    let steps_lenght = height * 0.41;
+
+    movement_from_gps(data)
+        .into_iter()
+        .filter_map(|this| {
+            if SPEED_THRESHOLD_KMPHR < this.speed_kmhr() {
```

Comment:
Please, add optional arg or context that will help to adjust treshold. Trained people could have something like 23-25 avarage kmphr while non trained people have 12-16 kmphr.